arch/x86/core/thread.c

```diff
@@ -175,93 +106,85 @@ __asm__("\t.globl _thread_entry\n"
- *
- *
- * @return opaque pointer to initialized k_thread structure
  */
 void _new_thread(struct k_thread *thread, k_thread_stack_t stack,
-		 size_t stackSize,
-		 _thread_entry_t pEntry,
+		 size_t stack_size, _thread_entry_t entry,
 		 void *parameter1, void *parameter2, void *parameter3,
 		 int priority, unsigned int options)
 {
-	char *pStackMem;
-
-	_ASSERT_VALID_PRIO(priority, pEntry);
-
-	unsigned long *pInitialThread;
+	char *stack_buf;
+	char *initial_esp;
+	struct _x86_initial_frame *initial_frame;
 
+	_ASSERT_VALID_PRIO(priority, entry);
 #if CONFIG_X86_STACK_PROTECTION
 	_x86_mmu_set_flags(stack, MMU_PAGE_SIZE, MMU_ENTRY_NOT_PRESENT,
 			   MMU_PTE_P_MASK);
 #endif
-	pStackMem = K_THREAD_STACK_BUFFER(stack);
-	_new_thread_init(thread, pStackMem, stackSize, priority, options);
-
-	/* carve the thread entry struct from the "base" of the stack */
-
-	pInitialThread =
-		(unsigned long *)STACK_ROUND_DOWN(pStackMem + stackSize);
-
-	/*
-	 * Create an initial context on the stack expected by the _Swap()
-	 * primitive.
-	 */
-
-	/* push arguments required by _thread_entry() */
-
-	*--pInitialThread = (unsigned long)parameter3;
-	*--pInitialThread = (unsigned long)parameter2;
-	*--pInitialThread = (unsigned long)parameter1;
-	*--pInitialThread = (unsigned long)pEntry;
-
-	/* push initial EFLAGS; only modify IF and IOPL bits */
-
-	*--pInitialThread = (EflagsGet() & ~EFLAGS_MASK) | EFLAGS_INITIAL;
-
+	stack_buf = K_THREAD_STACK_BUFFER(stack);
+	_new_thread_init(thread, stack_buf, stack_size, priority, options);
+
+	initial_esp = (char *)STACK_ROUND_DOWN(stack_buf + stack_size);
+
+	/* Create an initial context on the stack expected by _Swap() */
+	initial_frame = (struct _x86_initial_frame *)
+		(initial_esp - sizeof(struct _x86_initial_frame));
+	/* _thread_entry() arguments */
+	initial_frame->entry = entry;
+	initial_frame->p1 = parameter1;
+	initial_frame->p2 = parameter2;
+	initial_frame->p3 = parameter3;
+	/* initial EFLAGS; only modify IF and IOPL bits */
+	initial_frame->eflags = (EflagsGet() & ~EFLAGS_MASK) | EFLAGS_INITIAL;
 #if defined(CONFIG_GDB_INFO) || defined(CONFIG_DEBUG_INFO) \
 	|| defined(CONFIG_X86_IAMCU)
-	/*
-	 * Arrange for the _thread_entry_wrapper() function to be called
-	 * to adjust the stack before _thread_entry() is invoked.
-	 */
-
-	*--pInitialThread = (unsigned long)_thread_entry_wrapper;
-
-#else /* defined(CONFIG_GDB_INFO) || defined(CONFIG_DEBUG_INFO) */
-
-	*--pInitialThread = (unsigned long)_thread_entry;
-
-#endif /* defined(CONFIG_GDB_INFO) || defined(CONFIG_DEBUG_INFO) */
-
-	/*
-	 * note: stack area for edi, esi, ebx, ebp, and eax registers can be
-	 * left
-	 * uninitialized, since _thread_entry() doesn't care about the values
-	 * of these registers when it begins execution
-	 */
-
-	/*
-	 * The k_thread structure is located at the "low end" of memory set
-	 * aside for the thread's stack.
+	 /* Adjust the stack before _thread_entry() is invoked */
```

Comment:
Not relevant to this patch (except in the sense of being further cleanup), but if we have a abstracrted "wrapper" framework to handle these features that need to hook thread startup, then we should be predicating the use of that wrapper on a similarly abstract thing like NEED_THREAD_ENTRY_WRAPPER.

Or even have a function like get_thread_entry(_thread_entry) that does the wrapping appropriately at compile time as needed.

Reply:
a much later patch in my userspace series sorta does this.